dashboard/utils.py

```diff
@@ -0,0 +1,359 @@
+            cls._bind_map[h.__func__.__route_method__][
+                h.__func__.__route_path__].instance = master_instance
+
+
+def get_all_modules(module_type):
+    logger.info("Get all modules by type: {}".format(module_type.__name__))
+    import ray.new_dashboard.modules
+
+    for module_loader, name, ispkg in pkgutil.walk_packages(
+            ray.new_dashboard.modules.__path__,
+            ray.new_dashboard.modules.__name__ + "."):
+        importlib.import_module(name)
+    return module_type.__subclasses__()
+
+
+def to_posix_time(dt):
+    return (dt - datetime.datetime(1970, 1, 1)).total_seconds()
+
+
+async def rest_response(success, message, **kwargs) -> aiohttp.web.Response:
+    return aiohttp.web.json_response({
+        "result": success,
+        "msg": message,
+        "data": to_google_style(kwargs)
+    })
+
+
+def to_camel_case(snake_str):
+    """Convert a snake str to camel case."""
+    components = snake_str.split("_")
+    # We capitalize the first letter of each component except the first one
+    # with the 'title' method and join them together.
+    return components[0] + "".join(x.title() for x in components[1:])
+
+
+def to_google_style(d):
+    """Recursive convert all keys in dict to google style."""
+    new_dict = {}
+    for k, v in d.items():
+        if isinstance(v, dict):
+            new_dict[to_camel_case(k)] = to_google_style(v)
+        elif isinstance(v, list):
+            new_list = []
+            for i in v:
+                if isinstance(i, dict):
+                    new_list.append(to_google_style(i))
+                else:
+                    new_list.append(i)
+            new_dict[to_camel_case(k)] = new_list
+        else:
+            new_dict[to_camel_case(k)] = v
+    return new_dict
+
+
+def message_to_dict(message, decode_keys=None, **kwargs):
+    """Convert protobuf message to Python dict."""
+
+    def _decode_keys(d):
+        for k, v in d.items():
+            if isinstance(v, dict):
+                d[k] = _decode_keys(v)
+            if isinstance(v, list):
+                new_list = []
+                for i in v:
+                    if isinstance(i, dict):
+                        new_list.append(_decode_keys(i))
+                    else:
+                        new_list.append(i)
+                d[k] = new_list
+            else:
+                if k in decode_keys:
+                    d[k] = binary_to_hex(b64decode(v))
+                else:
+                    d[k] = v
+        return d
+
+    if decode_keys:
+        return _decode_keys(
+            MessageToDict(message, use_integers_for_enums=False, **kwargs))
+    else:
+        return MessageToDict(message, use_integers_for_enums=False, **kwargs)
+
+
+def redirect_stream(stdout_filename, stderr_filename):
+    """Redirect stdout or stderr to file."""
+    if stdout_filename:
+        fout = open(stdout_filename, "a")
+        os.dup2(fout.fileno(), sys.stdout.fileno())
```

Comment:
close the fout?

Reply:
The `fout` is the redirected fd of stdout, we can't close it. `os.dup2` will close the original fd of stdout.
Please refer to:
https://docs.python.org/3/library/os.html?highlight=dup2#os.dup2
https://man7.org/linux/man-pages/man2/dup.2.html

Reply:
Thanks! Good to know!